### DISCUSSION TITLE
GraphQL folks, I'm looking for feedback on my async graphql repo fastapi repo. Help anyone?

### DISCUSSION THREAD

**Discussion by matthewchung74** [fastapi/fastapi#2854] [Questions] [answered]
I had a hard time looking for a complete example of graphene / graphql with async, so I made an example. I'm looking for feedback if anyone can chime in. it is based on full-stack-fastapi-postgresql

https://github.com/matthewchung74/full-stack-fastapi-graphql/

It's very much a work in progress but all the graphene pytest work in regards to users and items. Anyways, any feedback is appreciated. Especially around the auth part using depends.

**Comment by ArcLightSlavik** [comment]
# Not really feedback just information

Starlette is deprecating It's GraphQL support https://github.com/encode/starlette/pull/1135 and as @Kludex said we are not against it, never worked with GraphQL so not sure how big of a deal it is, but be aware of it.

**Reply by matthewchung74** [reply]
ouch. but thanks for letting me know. in light of that, this project as it is doesn't really make sense.

**Comment by lightoyou** [comment]
You remove your project ?

**Reply by matthewchung74** [reply]
i made it private since i wasn't going forward with it. i made it public again for now if you'd like to fork.

**Comment by dmtrs** [comment]
Hey @matthewchung74 
 
I was not able to see your repository, here are my 0.2 cents though.

For me it does not make sense to blend fastapi with graphql, although you probably can. In my mind GraphQL is a "framework" that introduce a layer on top of http and let you not care at all about the http semantics and focus on you logic/domain model/schema, what you wanna call it. Regularly, you would have a single route (no routing), you won't care about http status codes or http methods. On the other side fastapi is a "framework" that empower to work the http semantics serve your logic by routing, responses with appropriate status codes etc.

What I am trying to say is that using a graphql on top of fastapi would not introduce any benefit rather gives you trouble potentially.

**Reply by matthewchung74** [reply]
@lightoyou whoops, i just realized, i deleted the repo and I don't have a local copy. sorry. i did start another version here using a different graphql version which has a working example of subscriptions https://github.com/matthewchung74/fastapi-graphql-starter but have decided fastapi and graphql doesn't make too much sense. 

many of the reasons are what @dmtrs said above.  i found myself working around and fighting the framework too much

**Comment by tshelver** [comment]
For me, the lack of graphql support is a bit of a deal breaker.  Looking for one API platform to support the entire backend.
Most of what we are serving is  PostgreSQL data, so I can use something like Postgraphile to directly serve that data via GraphQL.  
However, I still need to integrate other (REST) resources for the client apps.  The point of having a GraphQL API server is that it consolidates all information sources into a singe http point and allows for a single GarphQL query / mutation / subscription from client apps regardless of the original data sources.

I can theoretically extend Postgraphile with javascript plugins to cover additional sources, but am not really comfortable with JS backend programming, and already  have python code covering much of our solution space to integrate our data with other systems such as Telegram (via bots).

Just my POV

**Comment by lightoyou** [comment]
Do you think just use directly startlette and ariadne when you wan't do graphql backend is better option than use fastapi?

The real question is that, if fastapi not fill this need do you have another option to suggest ?

**Reply by matthewchung74** [reply]
@lightoyou "Do you think just use directly startlette and ariadne when you wan't do graphql backend is better option than use fastapi?" I do. 

"if fastapi not fill this need do you have another option to suggest" i hear hasura.io is good. this is a personal pref kind of thing. for me, i wanted to keep my stack open source to serve ml models, so fast API was my first option, but am now finding aws amplify, although closed source, begrudgingly is a better fit, since all that other stuff that goes around an API to build an app needs to be built too.

@tshelver hasura might be worth exploring as well.

**Reply by tshelver** [reply]
I have been working with Postgraphile for a while, it's an independent open source effort (with paid subs for support and so on).  Happy enough with how Postgraphile works.  My only drawback with it is that my javascript is not great in case I need to create extensions.
However the maintainer is extremely helpful and very responsive, and there is a lot of sample code around creating extensions, subscriptions and live feeds, to name just a few.
He also has a lot of 'best practices' in terms of using Postgresql as a GraphQL backend which make a lot of sense, for those who are not totally familiar with Postgres security and access models, schemas, stored functions / procedures and so on.

Also quite a reasonable extension ecosystem already in place.

Have already used Postgraphile to do things like pull images stored in Postgres via GraphQL,  vehicle position and tracks and so on.  

Both products basically front end PostgreSQL databases and spin up GraphQL schemas

Where I was looking for is accessing information outside of Postgres. for example doing reverse geocoding to pull addresses from Google Maps REST interface or similar. 
I have another requirement around that that demands some fairly close integration with our Postgres PostGIS-based storage.
I can do this either via the mentioned extension, or Postgres also has an http extension which could possible be used, or could access from dart, then feed back into the API.

Generally I would not use either hasura or postgraphile for the classic GraphQL server API creating,  where it integrates several end data points into one schema.  That needs a different solution that is designed for that use case.  If FastAPI had good graphql support, it could work well, but...

**Comment by tiangolo** [comment, marked as answer]
If you really want to work with GraphQL, you might want to check out Strawberry: https://fastapi.tiangolo.com/advanced/graphql/

**Reply by acidjunk** [reply]
Strawberry works/integrates really nice with FastAPI. It also relies heavily on types.
